extensions/INTEL/SPV_INTEL_fpga_loop_controls.asciidoc

```diff
@@ -0,0 +1,136 @@
+SPV_INTEL_fpga_loop_controls
+============================
+
+== Name Strings
+
+SPV_INTEL_fpga_loop_controls
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/KhronosGroup/SPIRV-Headers
+
+== Contributors
+
+- Joe Garvey, Intel +
+- Michael Kinsner, Intel
+
+== Notice
+
+Copyright (c) 2019 Intel Corporation.  All rights reserved.
+
+== Status
+
+Final draft
+
+== Version
+
+[width="40%",cols="25,25"]
+|========================================
+| Last Modified Date | {docdate}
+| Revision           | A
+|========================================
+
+== Dependencies
+
+This extension is written against the SPIR-V Specification,
+Version 1.3 Revision 6.
+
+This extension requires SPIR-V 1.0.
+
+== Overview
+
+This extension adds a mechanism to allow more than the current limit of 32 loop controls.  It uses that new mechanism to add new loop controls for FPGA targets.
```

Comment:
Because one of those bits could be used as a continuation bit to the next word, it's not quite accurate to say there is a limit of only 32 controls.

The new controls proposed do seem in good company with some existing controls, so I think Khronos should look at reserving some existing bits for this purpose.

Reply:
Hi John,

I've updated the overview wording through an additional commit.

We don't disagree with your comments.  Are you asking for one of the following options up front, or just making an observation with your comment?

1. We can make some of these controls use bitfield locations in the future, once there is a way to allocate bits for specific extensions.  I'd expect a future revision of this extension to factor in that change.

2. We could split the extension into two parts, one that defines extended loop controls trailing as additional operands, and one that defines the additional loop controls.  We need the additional operands for some controls (bitfield isn't sufficient) because there can be repeated instances of certain controls on the same OpLoopMerge - e.g. DependencyArrayINTEL.  Our thought is that Khronos should define the 0x80000000 bitfield longer term (matching the extension functionality), and then we can remove that part from this extension.

I think two bitfield values need to be formally defined longer term: (1) what we've defined here as 0x80000000, where additional operands define loop controls which may repeat; (2) the extended bitfield case, indicating that the following word is an additional bitfield for loop controls.

Otherwise, please let me know what we should do in order for this to be merged.  Are you suggesting that we should switch to formal bitfield allocations for some of these controls (the non-repeating ones) before merging the extension to the registry?

Reply:
Intel has the final decision of what goes in the registry within an Intel vendor-specific extension. I just wanted to leave this open long enough to first gather feedback and have some discussion. This can potentially avoid greater pain further down the road.

Reply:
The extension has been updated to use the loop control bitfield allocation mechanism now in SPIRV-Headers/spir-v.xml.  Much cleaner now!